### PR TITLE
Fix OTA E2E test failures

### PR DESCRIPTION
**Pull request by tgsong**
Description
-----------

- Directly return from callback if invoked after http backend stopped.
- Never call http connect in callback since the OTA agent thread may stop the data plane.
- Disconnect from S3 when OTA agent cleans up the http data plane.
- Cancel http request if http downloader is already stopped.
- Only free OTA http buffer when OTA agent is stopping.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c [hunk 1]
```diff
@@ -1177,7 +1185,15 @@ OTA_Err_t _AwsIotOTA_CleanupData_HTTP( OTA_AgentContext_t * pAgentCtx )
     /* Unused parameters. */
     ( void ) pAgentCtx;
 
-    memset( &_httpDownloader, 0, sizeof( _httpDownloader_t ) );
+    /* Disconnect from the S3 server. */
+    IotHttpsClient_Disconnect( _httpDownloader.httpConnection.connectionHandle );
```
Comment (sarenameas) [inline libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c:1189]: For possible warnings?
```suggestion
    ( void ) IotHttpsClient_Disconnect( _httpDownloader.httpConnection.connectionHandle );
```

Comment (pvyawaha) [inline libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c:1189]: Can we log the return code here for debug log ?

File: libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c [hunk 2]
```diff
@@ -435,6 +436,12 @@ static void _httpReadReadyCallback( void * pPrivateData,
     /* Buffer to read the "Connection" field in HTTP header. */
     char connectionValueStr[ HTTP_HEADER_CONNECTION_VALUE_MAX_LEN ] = { 0 };
 
+    /* Bail out if this callback is invoked after http downloader stopped. */
+    if( _httpDownloader.state == OTA_HTTP_STOPPED )
+    {
+        return;
```
Comment (sarenameas) [inline libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c:442]:
```suggestion
        IotHttpsClient_CancelResponseAsync( responseHandle );
        return;
```

Comment (sarenameas) [reply libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c:442]: But this will invoke the errorcallback saying saying the user cancelled.

Comment (sarenameas) [reply libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c:442]: Per offline conversation `IotHttpsClient_CancelRequestAsync( requestHandle );` will also be added to _httpAppendHeaderCallback().

**Review by sarenameas — DISMISSED** [review]

**Review by sarenameas — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Pull request merged**